internal/provider/resource_gitlab_project_environment.go

```diff
@@ -0,0 +1,175 @@
+package provider
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"net/http"
+	"strconv"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	gitlab "github.com/xanzy/go-gitlab"
+)
+
+// https://docs.gitlab.com/ee/ci/environments/protected_environments.html
+var _ = registerResource("gitlab_project_environment", func() *schema.Resource {
+	return &schema.Resource{
+		Description: "This resource allows you to create and manage GitLab environments.",
+
+		CreateContext: resourceGitlabProjectEnvironmentCreate,
+		ReadContext:   resourceGitlabProjectEnvironmentRead,
+		UpdateContext: resourceGitlabProjectEnvironmentUpdate,
+		DeleteContext: resourceGitlabProjectEnvironmentDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+		Schema: map[string]*schema.Schema{
```

Comment:
I've just checked are there are actually already some more fields available:

https://github.com/xanzy/go-gitlab/blob/bbf334f1f1635442ca8804b919a21bf444c78940/environments.go#L35

* `slug`
* `last_deployment`

What's missing in go-gitlab (and I'm going to add there) is:

* `created_at`
* `updated_at`

... these are all `Computed` fields btw.

Reply:
see https://github.com/xanzy/go-gitlab/pull/1414

Reply:
However, these two fields are no show stopper for this PR, just at least implement the `slug` attribute.

Reply:
I see your pull request was merged into go-gitlab I can add a commented out version for now of `created_at` and `updated_at` for now until go-gitlab is released with that MR. `slug` should be easy its `last_deployment` I worry about. The  attribute might a diff every single time someone deploys to the environment which would be annoying to look at in a terraform plan/apply

Reply:
Let's not worry about `last_deployment` for now, but `slug` and if go-gitlab is released in time, we can also implement `created_at` and `updated_at` - otherwise, we'll do this in the next increment.

Reply:
I have unresolved this thread because I see that the latest go-gitlab was released with your changes for `created_at` and `updated_at`.  When I have correctly added these please resolve this thread again.  Also is there a way/point to unit testing these 2 computed fields? Since GitLab generates these 2 fields internally we would have to fuzz the functional testing logic to account for differences in time.

Reply:
I've added some suggestions:

1. writing them to state
2. testing these two fields by using what is returned from `testAccCheckGitlabProjectEnvironmentExists()` and testing against the state.

I'm resolving this thread for now.